airflow/config_templates/default_airflow.cfg

```diff
@@ -248,6 +248,16 @@ dag_processor_manager_log_location = {AIRFLOW_HOME}/logs/dag_processor_manager/d
 # Default to use task handler.
 task_log_reader = task
 
+[creds_backend]
+# List of creds backend classes to use in order of desired search path
+class_list = airflow.creds.environment_variables.EnvironmentVariablesCredsBackend, airflow.creds.metastore.MetastoreCredsBackend
+
+# Use when using AWS SSM / airflow.providers.amazon.aws.creds.ssm.AwsSsmCredsBackend as creds provider
+# If prefix is ``/airflow``, then example param path is ``/airflow/AIRFLOW_CONN_SMTP_DEFAULT``
+# Use same URI format as you would with env vars, or ``conn_uri`` CLI param.
+aws_ssm_prefix = /airflow
+aws_ssm_profile_name =
```

Comment:
I think we should have `creds_backend_kwargs` or something similar to avoid harcoding configs related to a specific provider. WDYT?

If we don't do that we would need to hardcode this configs when I (or someone) adds support for Google's KMS and Hashicorp Vault :)

Reply:
ok so (1) the other thought i had was basically this could be a one line config `creds_backend_config` under `[core]`.  then it would have a config like this (but one-line-ified):
```
{
"airflow.providers.amazon.aws.creds.ssm.AwsSsmCredsBackend": {"profile_name": "default", "ssm_prefix": "/my_airflow_creds"},
"airflow.creds.environment_variables.EnvironmentVariablesCredsBackend": null, 
"airflow.creds.metastore.MetastoreCredsBackend": null
}
```
so this would determine the search path _and_ provide all config options.

is that kinda what you are thinking?

or (2) are you just thinking that it is like this:
```
[creds_backend]
class_list = blah.class1, blah.class2  # this is the search path
creds_backend_kwargs = {"this": "will", "be": "passed", "to": "all": "backends"}
```

the problem with (2), is you can't specify which kwargs set will go to which backend (since we can support multiple simultaneously)

the problem with (1) is it seems a bit unwieldy and more fiddly to work with so i thought it better to just use INI params with a prefix to "namespace" each backend.

would you be happy with adding `google_kms_kwargs` under `[creds_backend]`.  

other ideas?

Reply:
What I think and this is just my opinion is that we should simplify it as follows:

1) We don't have a search path and instead have a fixed Precedence and we support only 1 External Cred Backend
2) The flow would be if the class_path for External Cred Backend is not None, a) use that, fallback to b) Environment Variable if connectin is not available in External Cred Backend and then c) Connection in Metadata DB

Although this limits the feature as you won't be able to use multiple CredBackends but would we ever want to use multiple external backends ( I don't know)? I can definitely imagine the need of using more than 1 `ssm_prefix` but not sure of the former. 

Would love to hear your opinion.

~Regarding the same topic, how would we support multiple ssm_prefix? i.e. let's say we have some secrets in `/my_airflow_creds_data_scientists` and `/my_airflow_creds_devops`?~  hmm.. maybe we don't need to support multiple ssm_prefix too 🤔 , Sysadmin/Devops can create a dedicated prefix for Airflow

Reply:
i am totally fine with this idea.  makes configuration simpler and more generic.   

the only sticking point would be if someone thinks we really should support multiple backends, OR if someone thinks it's important to customize search path.   

i guess i'll let this comment linger for a day-ish and barring an objection make the change.

re multiple ssm prefix... to support this we could do `aws_ssm_prefix_list` instead... but... like you indicated, probably not necessary?

thanks :)

Reply:
update: i have renamed creds --> secrets

i think we have general consensus that secrets is a better name

the other idea about config structure (and removing configurability of search path), i am going to continue to let that percolate a little bit to see if there are any other opinions about it.  i have also updated the AIP to reflect the new provisional name

Reply:
imo having a configurable search path is useful. We're in the middle of migrating secrets from various config files to Vault, and having the ability to query multiple backends will ease the transition. Add to that migrating our current airflow secrets (where some are in the env and some are in the airflow meta-db) to a Vault secrets backend, having a search path where we can just put Vault at the front will make the change way easier.

that said, it would be possible to implement a custom secrets backend that just searches a list of other secrets backends specific to our requirements.

Reply:
> that said, it would be possible to implement a custom secrets backend that just searches a list of other secrets backends specific to our requirements.

yeah because of how it would be pretty straightforward to do this kind of thing, for me this tips the scale over to just allowing one, because doing so makes the config simpler.  the config would then just need a single item like `creds_backend_kwargs` -- no need for configuration of search path, and no need to namespace the individual backend-specific configuration, which would be required for supporting multiple simultaneously.